### PR TITLE
[AIRFLOW-689] Okta Authentication

### PR DESCRIPTION
**Pull request by byang2013**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-689

Testing Done:
1. Install [Okta SDK](http://developer.okta.com/docs/sdk/core/python_api_sdk/) with pip onto your airflow instance
2. Add in your [Okta API key](http://developer.okta.com/docs/api/getting_started/getting_a_token.html) and organization URL (usually your_org.okta.com) in config
3. Replace backend with okta_auth in config
4. Log in

### CODE REVIEW

**Comment by bowenli86** [comment]
The landscape thing is very weird. I ran into it when submitting [my PR](https://github.com/apache/incubator-airflow/pull/1926).

Airflow committers'd better remove that check, and add it back after landscape.io fixes the bug

File: airflow/contrib/auth/backends/okta_auth.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        username = request.form.get("username")
+        password = request.form.get("password")
+
+    if not username or not password:
+        return self.render('airflow/login.html',
+                           title="Airflow - Login",
+                           form=form)
+
+    url = configuration.get("okta", "url")
+    key = configuration.get("okta", "api_key")
```
Comment (msumit) [inline airflow/contrib/auth/backends/okta_auth.py:105]: instead of configuration, it would be better if we pick values from connections (db).

**Comment by msumit** [comment]
There are dozen's of saml providers and okta is just one of them... so instead of `okta_auth` there should be `saml_auth`, using that user's should be able to configure any saml provider.

File: airflow/contrib/auth/backends/okta_auth.py (hunk 1 above)
Comment (bolkedebruin) [reply airflow/contrib/auth/backends/okta_auth.py:105]: @msumit that doesn't make sense, this is not a user exposed connection.

Comment (msumit) [reply airflow/contrib/auth/backends/okta_auth.py:105]: yeah, make sense..

**Comment by argh128** [comment]
@msumit The code as written is not a SAML login method; but instead just off loads the authentication to Okta (kind of like ldap/ad/radius auth checks). This code would not be re-usable by other SAML providers;

(Not the author, not a member of the airflow project, just reviewing the commit)

**Comment by msumit** [comment]
@argh128 that's not completely correct.. In general (if not all) saml providers works on a `custom_url` and `api_key` concept, given these 2 values and any provider would be able to do saml authorization for you..

**Comment by argh128** [comment]
My comment is more that this is not generalizable, and is not using a saml assertion, so trying to call it saml_auth or having it be configurable for multiple providers is not really possible with the way this PR is constructed. It does do what Bryan requested which is a simple way to route the authentication to Okta. 

IMO, I would use the sample code from OneLogin https://github.com/onelogin/python-saml/blob/master/demo-flask/index.py as a base which does use saml, and thus could be generalized to multiple providers. 

The code as written is very specific to Okta, because Okta provides an authentication API that allows you to check if a username/password matches the okta password credentials. It would not work in our environment because we require 2FA which would then require a second factor check that further makes this Okta specific.

**Comment by argh128** [comment]
I should also mention that at least onelogin also provides a mechanism like the one your thinking for auth https://developers.onelogin.com/api-docs/1/users/create-session-login-token

But besides it using a token, password and username has little resemblance to the okta one.

**Comment by andylockran** [comment]
This isn't working at the moment for me - but would love to help get this fixed and into prod.  I get 405 method not allowed when trying to login.

@byang2013 are you still interested in getting this merged in?

**Comment by someword** [comment]
Bummer this has not been merged.   I tested with airflow 1.9 and it works great.

**Comment by Fokko** [comment]
@byang2013 Can you:

- rebase against master
- squash the commits to a single commit
- Prepend `[AIRFLOW-689]` to the commit.

**Comment by someword** [comment]
@Fokko - The PR author has not responded to previous queries.  Is it possible someone else can adopt and do what's needed to the PR to get it merged?

**Comment by Fokko** [comment]
@someword Are you using this? The code is old and untested.

**Comment by someword** [comment]
@Fokko - i am using it and it works great.

**Comment by opragel** [comment]
To me a generic SAML integration, or a generic OAuth/OpenID Connect authentication method makes more sense than dealing with Okta's API directly. The Okta API tokens are a bit dangerous access-wise. Airflow also already has GitHub and Google OAuth integrations, and Okta supports OpenID & OAuth, so that might be the most natural thing to add.

https://airflow.apache.org/security.html#oauth-authentication

**Comment by jimsmith** [comment]
+1 we rely on Okta for SSO and this is extremely useful for our users and business needs, please merge this in.

**Comment by r39132** [comment]
@byang2013 Are you interested in taking this forward? If I don't hear back in a few days, I'll close this PR. @someword @jimsmith @msumit @andylockran if one of you would like to take this word forward in a new PR, please do so.

**Comment by r39132** [comment]
Folks! I'm closing this. If someone wants to take this forward, please open a new PR.

**Comment by maxg203** [comment]
I'd be interested in adopting this Okta integration work to get it merged if that is still a possibility. I think an Okta authentication backend would be useful to a lot of people.

**Comment by ashb** [comment]
The new RBAC-based Webserver is what we recommend people use going forward, and to integrate Okta with that something like https://github.com/ministryofjustice/fab-oidc is what I would recommend.

Many providers support OIDC, Okta included.

**Comment by msumit** [comment]
Folks who were interested in Okta's api based authentication, I've raised another PR (https://github.com/apache/incubator-airflow/pull/4143). Give it a try and see how it goes for you. 

@ashb agrees with your thought, but folks who want simplicity, there is no harm in having one more auth provider in the contrib section.

**Pull request closed without merging**